File: ThirtyFour.UI/ThirtyFour.UI.Tests/Table/ColumnHeaderCellFinderTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using ThirtyFour.UI.Table;
using System;

namespace ThirtyFour.UI.Tests.Tests.Table
{
    [TestClass]
    public class ColumnHeaderCellFinderTests : BaseTestSuite
    {

        [TestMethod]
        public void InavlidFindColumnHeaderThrowsException()
        {
            string invalidColumn = "shdgydbhdiu";
            ColumnHeaderCellFinder finder = new ColumnHeaderCellFinder(invalidColumn, "First Test Product", "Subtotal");
            try
            {
                IWebElement tableElement = this.driver.FindElement(By.ClassName("cart"));
                finder.FindCell(tableElement);
            }
            catch (NoSuchElementException ex)
            {
                string message = String.Format("Unable to find column with header text '{0}'", invalidColumn);
                Assert.AreEqual(message, ex.Message);
            }
        }

        [TestMethod]
        public void InavlidReturnColumnHeaderThrowsException()
        {
            string invalidColumn = "shdgydbhdiu";
            ColumnHeaderCellFinder finder = new ColumnHeaderCellFinder("Product", "First Test Product", invalidColumn);
            try
            {
                IWebElement tableElement = this.driver.FindElement(By.ClassName("cart"));
                finder.FindCell(tableElement);
            }
            catch (NoSuchElementException ex)
            {
                string message = String.Format("Unable to find column with header text '{0}'", invalidColumn);
                Assert.AreEqual(message, ex.Message);
            }
        }

        [TestMethod]
        public void InavlidFindValueThrowsException()
        {
            string invalidValue = "shdgydbhdiu";
            ColumnHeaderCellFinder finder = new ColumnHeaderCellFinder("Product", invalidValue, "Subtotal");
            try
            {
                IWebElement tableElement = this.driver.FindElement(By.ClassName("cart"));
                finder.FindCell(tableElement);
            }
            catch (NoSuchElementException ex)
            {
                string message = String.Format("Unable to find row containing find value '{0}'", invalidValue);
                Assert.AreEqual(message, ex.Message);
            }
        }

        [TestMethod]
        public void TestFindCellReturnsCorrectWebElement()
        {
            ColumnHeaderCellFinder finder = new ColumnHeaderCellFinder("Product", "First Test Product", "Subtotal");
            IWebElement tableElement = this.driver.FindElement(By.ClassName("cart"));
            IWebElement cell = finder.FindCell(tableElement);
            Assert.AreEqual("$9.99", cell.Text);
        }
    }
}

File: java/src/test/java/com/github/danhumphrey/thirtyfour/ui/test/table/ColumnHeaderCellFinderTests.java
package com.github.danhumphrey.thirtyfour.ui.test.table;

import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import com.github.danhumphrey.thirtyfour.ui.table.ColumnHeaderCellFinder;
import com.github.danhumphrey.thirtyfour.ui.test.BaseTestSuite;

import static org.hamcrest.core.StringStartsWith.startsWith;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;

public class ColumnHeaderCellFinderTests extends BaseTestSuite {
	
	@Test
	public void testInvalidFindColumnHeaderThrowsNoSuchElementException() {
		WebDriver driver = getDriver();
		String invalidColumn = "dudhuidhiudh";
		ColumnHeaderCellFinder finder = new ColumnHeaderCellFinder(invalidColumn, "First Test Product", "Subtotal");
		try {
			WebElement tableElement = driver.findElement(By.className("cart"));
			finder.findCell(tableElement);
		}catch(NoSuchElementException ex) {
			String exMessage ="Unable to find column with header text " + invalidColumn;
			assertThat(ex.getMessage(),  startsWith(exMessage));
			return;
		}
		fail("Expected NoSuchElementException to be thrown when we provide an invalid find column header");
		
	}
	
	@Test
	public void testInvalidReturnColumnHeaderThrowsNoSuchElementException() {
		WebDriver driver = getDriver();
		String invalidColumn = "dudhuidhiudh";
		
		ColumnHeaderCellFinder finder = new ColumnHeaderCellFinder("Product", "First Test Product", invalidColumn);
		try {
			WebElement tableElement = driver.findElement(By.className("cart"));
			finder.findCell(tableElement);
		}catch(NoSuchElementException ex) {
			String exMessage = "Unable to find column with header text " + invalidColumn;
			assertThat(ex.getMessage(), startsWith(exMessage));
			return;
		}
		fail("Expected NoSuchElementException to be thrown when we provide an invalid return column header");
		
	}
	
	@Test
	public void testInvalidFindValueThrowsNoSuchElementException() {
		WebDriver driver = getDriver();
		String invalidValue = "dudhuidhiudh";
		
		ColumnHeaderCellFinder finder = new ColumnHeaderCellFinder("Product", invalidValue, "Subtotal");
		try {
			WebElement tableElement = driver.findElement(By.className("cart"));
			finder.findCell(tableElement);
		}catch(NoSuchElementException ex) {
			String exMessage = "Unable to find cell with text " + invalidValue;
			assertThat(ex.getMessage(), startsWith(exMessage));
			return;
		}
		fail("Expected NoSuchElementException to be thrown when we provide an invalid find value");
	}
	
	@Test
	public void testFindCellReturnsCorrectWebElement() {
		WebDriver driver = getDriver();
		
		ColumnHeaderCellFinder finder = new ColumnHeaderCellFinder("Product", "First Test Product", "Subtotal");
		WebElement tableElement = driver.findElement(By.className("cart"));
		WebElement cell = finder.findCell(tableElement);
	
		assertThat(cell.getText(), equalTo("$9.99"));
	}
}
